### PR TITLE
ToMap Command

### PR DESCRIPTION
**Pull request by mvrozanti**
Proof of concept implementation of issue #2247

### CODE REVIEW

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                    }
+                } catch (IOException e) {
+                    System.out.println("Error setting color for " + blockId);
+                    e.printStackTrace();
+                }
+            }
+        }
+        logDirect("Finished coloring");
+        try {
+            ImageIO.write(result, "png", new File("/home/nexor/output.png"));
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:136]: don't hardcode output directory
Better save it to screenshots or the world's baritone folder

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public Color averageColor(BufferedImage bi, String id) {
+        if (!cachedColors.containsKey(id)) {
+//            System.out.println(id);
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:144]: don't leave debug output as comments

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                cachedColors.put(id, new Color(sumr / num, sumg / num, sumb / num));
+            } catch (Exception e) {
+                cachedColors.put(id, new Color(Color.TRANSLUCENT));
+            }
+        }
+        return cachedColors.get(id);
+    }
+
+    private static BufferedImage getTextureForBlockId(String blockId) throws IOException {
+        String base = "/home/nexor/.minecraft/versions/1.15.2-OptiFine_HD_U_G5_pre1/1.15.2-OptiFine_HD_U_G5_pre1/assets/minecraft/textures/block/";
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:167]: another hardcoded path

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                });
+            }
+        };
+    }
+
+    // http://jens-na.github.io/2013/11/06/java-how-to-concat-buffered-images/
+    // /home/nexor/.minecraft/versions/1.15.2-OptiFine_HD_U_G5_pre1/1.15.2-OptiFine_HD_U_G5_pre1/assets/minecraft/textures/block
+    private void concatImages() {
+
+    }
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:207]: why is there an empty method?

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    @Override
+    public Stream<String> tabComplete(String label, IArgConsumer args) throws CommandException {
+        return Stream.empty();
+    }
+
+    @Override
+    public String getShortDesc() {
+        return "null";
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:216]: maybe put something more informative here

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Override
+    public List<String> getLongDesc() {
+        return Arrays.asList("null");
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:221]: same here

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+public class ToMapCommand extends Command {
+
+    private HashMap<String, Color> cachedColors;
+
+    public ToMapCommand(IBaritone baritone, String... names) {
+        super(baritone, "tomap");
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:52]: either forward the names parameter or better remove it but don't leave it unused

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        cachedColors = new HashMap<>();
+    }
+
+    @Override
+    public void execute(String label, IArgConsumer args) throws CommandException {
+        ctx.worldData().getCachedWorld().reloadAllFromDisk();
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:58]: Shouldn't it as well repack everything it can repack so the data is more up to date?

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+                    String id;
+                    if (bs == null) {
+                        id = "air";
+                    } else {
+                        try {
+                            Matcher m = p.matcher(bs.toString());
+                            if (m.find())
+                                id = m.group(1);
+                            else
+                                throw new Exception("Damn");
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:81]: Maybe give more information on what caused the exception

**Review by ZacSharp — CHANGES_REQUESTED** [review]

**Comment by ZacSharp** [comment]
I know this is a proof of concept but you should iron out at least the hardcoded paths so others can test it

**Comment by mvrozanti** [comment]
I don't know how to take them out and turn them generally working

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 1 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:136]: How?

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 2 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:144]: This is non-important.

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 3 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:167]: How can we fix this?

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 4 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:207]: Just so I remembered that I need to do that at some point

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 5 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:216]: The issue is informative enough for now

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 6 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:221]: Same above

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 7 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:52]: Agreed, will fix in the future

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 8 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:58]: I don't know what repack means

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 9 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:81]: Agreed

**Comment by mvrozanti** [comment]
Can we focus on real problems on the issue assigned to this PR plz?

**Comment by ZacSharp** [comment]
`net.minecraft.client.Minecraft.getMinecraft().gameDir.toPath().resolve("screenshots")` for the screenshots directory and check `baritone.cache.WorldProvider#initWorld` for how to find the place where world/server specific baritone information is stored

Ok, what problems to you mean?

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 1 above)
Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:136]: https://github.com/cabaletta/baritone/pull/2257#issuecomment-743911392

**Comment by mvrozanti** [comment]
> net.minecraft.client.Minecraft.getMinecraft().gameDir.toPath().resolve("screenshots") for the screenshots directory and check baritone.cache.WorldProvider#initWorld for how to find the place where world/server specific baritone information is stored

Thanks I'm gonna work on that.

> Ok, what problems to you mean?
I mean mostly two things: 
1) how do we stitch the regions together? 
2) how do we efficiently serialize the image?

Given the code we already got.

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 3 above)
Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:167]: Don't know if https://github.com/cabaletta/baritone/pull/2257#issuecomment-743911392 is enough here but I think it should be possible to find the current version's assets

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 5 above)
Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:216]: This is the ingame help, for those that didn't read the issue (hopefully that is everyone soon)
I personally think it's better to do stuff like this before forgetting it, but maybe you are better in remembering stuff than I am

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 8 above)
Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:58]: It takes the chunks in render distance and caches them, so basically whenever you use the cache and have enough time, do a repack so you don't let the cache fool you

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 5 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:216]: I didn't even stop to think about a description. I think this should be the last thing to be written to this file.

**Comment by mvrozanti** [comment]
> `net.minecraft.client.Minecraft.getMinecraft().gameDir.toPath().resolve("screenshots")` for the screenshots directory and check `baritone.cache.WorldProvider#initWorld` for how to find the place where world/server specific baritone information is stored
> 
> Ok, what problems to you mean?

image

I'm missing this method.

**Comment by ZacSharp** [comment]
Just noticed you based this on 1.15. usually pulls are based on master, so searched there and maybe this has changed from 1.12 to 1.15

**Comment by ZacSharp** [comment]
It is `net.minecraft.client.Minecraft.getInstance().gameDir.toPath().resolve("screenshots")` now

**Comment by l1ving** [comment]
Yes, pulls should be done to master and backported upwards

**Comment by mvrozanti** [comment]
I'll work on these reviews but knowing there's still a long ways to go before merging to master

File: src/main/java/baritone/command/defaults/ToMapCommand.java (hunk 3 above)
Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:167]: The latter would be best, since not everyone has an unzipped jar lying around in their versions directory lol

**Comment by mvrozanti** [comment]
I'm getting some very weird behavior. Only region x=0,z=0 doesn't return null, even though reloadAllFromDisk is being called and there are countless other region files on disk.

**Comment by mvrozanti** [comment]
I'm getting some very weird behavior with this call:
`ICachedRegion cr = ctx.worldData().getCachedWorld().getRegion(rX, rZ);`

Only region x=0,z=0 doesn't return null, even though reloadAllFromDisk is being called and there are countless other region files on disk. I'm kinda stuck.

EDIT:

What's more weird, is that I get the following messages in the log after calling reloadAllFromDisk(): 

```
[00:31:37] [main/INFO]: [STDOUT]: Loading region 0,0 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache
[00:31:38] [main/INFO]: [STDOUT]: Loaded region successfully in 317ms
[00:31:38] [main/INFO]: [STDOUT]: Loading region 0,-1 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache
[00:31:38] [main/INFO]: [STDOUT]: Loaded region successfully in 303ms
[00:31:38] [main/INFO]: [STDOUT]: Loading region -20,46 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache
[00:31:38] [main/INFO]: [STDOUT]: Loaded region successfully in 48ms
[00:31:38] [main/INFO]: [STDOUT]: Loading region -21,46 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache
[00:31:38] [main/INFO]: [STDOUT]: Loaded region successfully in 24ms
[00:31:38] [main/INFO]: [STDOUT]: Loading region -1,0 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache
[00:31:38] [main/INFO]: [STDOUT]: Loaded region successfully in 314ms
[00:31:38] [main/INFO]: [STDOUT]: Loading region -1,-1 from disk /home/nexor/.minecraft/baritone/<my-server-ip-address>/DIM0/cache


```

So, not only getRegion is returning null for regions that have been (apparently) loaded succesfully, not all regions stored on disk are being read succesfully. What am I doing wrong here?

EDIT2: I must add, `~/.minecraft/baritone/<my-server-ip-address>/` is a symbolic link.

**Comment by ZacSharp** [comment]
I think `reloadAllFromDisc` reloads all regions that are already loaded and `tryLoadFromDisk` or `getOrCreateRegion` is more what you want.

**Comment by mvrozanti** [comment]
Cool let me try that

**Comment by mvrozanti** [comment]
image
Colors actually work now.

**Comment by mvrozanti** [comment]
Colors are now fairly consistent, thanks @ZacSharp .

But `tryLoadFromDisk` isn't exactly working... I need to test further but I think it's returning an empty CachedRegion.

**Comment by ZacSharp** [comment]
tryLoadFromDisk's return type is void

**Comment by mvrozanti** [comment]
> tryLoadFromDisk's return type is void

Yes. What do you mean?

**Comment by ZacSharp** [comment]
You said you think it returns an empty region, so I (indirectly) said it returns nothing

**Comment by mvrozanti** [comment]
I edited my previous message with what I actually meant.

**Comment by mvrozanti** [comment]
I have no idea how to proceed.

**Comment by ZacSharp** [comment]
And I can't test it because my hard drive is full

**Comment by mvrozanti** [comment]
Well, we tried.

**Comment by ZacSharp** [comment]
I haven't given up yet, it's just a bit more difficult

**Comment by mvrozanti** [comment]
I kinda have. There's no input from the repository admins and baritone should definitely have its bugs fixed before prioritizing new features.

**Comment by ZacSharp** [comment]
Can you please rebase this onto master so I can test it without gradle needing to cache new stuff?

**Comment by mvrozanti** [comment]
[Sure](https://github.com/mvrozanti/baritone/tree/tomap-command-from-master). You're definitely going to have to tweak a few things, such as the retrieving the screenshots directory.

**Comment by scorbett123** [comment]
I am not entirely sure, however I believe that there is a setting about whether to save cached chunks to the disk,  have you enabled it?

**Comment by mvrozanti** [comment]
@scorbett123 AFAIK, no. But I have 2GB of data under `~/.minecraft/baritone/<my-server-ip>`. I think that's enough proof that caching to disk is happening. But I can't get what's cached to disk to runtime memory.

**Comment by mvrozanti** [comment]
Funny thing is, somehow region 0,0 was loaded to runtime memory. I'm still not sure why it worked for this region and not others. The result is pretty neat:

**Comment by scorbett123** [comment]
Sorry, I wrote that late last night, was thinking of the wrong project.

**Comment by scorbett123** [comment]
Wait, why are you trying to merge into the 1.15.2 branch?

**Comment by mvrozanti** [comment]
> Wait, why are you trying to merge into the 1.15.2 branch?

I wasn't trying to merge yet, I just made a PR for convenience for people to give feedback. Also, I'm testing for 1.15.2 only

**Comment by ZacSharp** [comment]
Was the rebased branch outputting an image before the rebase? Just so I know what I screwed up by replacing the image loading and what was already broken.

**Comment by mvrozanti** [comment]
No, it wasn't. I actually didn't even try to run it.

**Comment by ZacSharp** [comment]
Not even before the rebase?

**Comment by mvrozanti** [comment]
I edited my answer.

**Comment by ZacSharp** [comment]
And I deleted my comment

**Comment by mvrozanti** [comment]
I'd advise to test this on 1.15.2 though. I've only actually tested in that version.

**Comment by ZacSharp** [comment]
Differences are only 5 replacements, so they should be negligible.
Bigger problem is that I don't have an assets folder somewhere and I'm testing in singleplayer.
I managed to get 6 transparent images now

**Comment by mvrozanti** [comment]
>  Bigger problem is that I don't have an assets folder somewhere and I'm testing in singleplayer.
Just unzip the jar for the version you're running located in `.minecraft/versions` and edit the string in `ToMapCommand#getTextureForBlockId`

**Comment by ZacSharp** [comment]
That one stings a bit but I don't think it is important right now: `Failed to find block id for minecraft:air`

**Comment by mvrozanti** [comment]
I don't think you have the right version. Can you please, please test this branch? We're making this waaaaaaaay harder than it has to be.

**Comment by ZacSharp** [comment]
This has to be ported to 1.12.2 at some point anyway, so why not do it now?

**Comment by mvrozanti** [comment]
Because it barely works on this version!

**Comment by mvrozanti** [comment]
> Failed to find block id for minecraft:air

If you're getting this message, you are most certainly not using the latest version of the code I wrote, or the rebased one.

**Comment by ZacSharp** [comment]
You changed something since https://github.com/cabaletta/baritone/pull/2257#issuecomment-744074936?

**Comment by mvrozanti** [comment]
I don't know man. Can't you check the git log or just download the latest?

**Comment by ZacSharp** [comment]
`git pull` outputs "up to date" for both branches

**Comment by mvrozanti** [comment]
This is getting out of hand..... Please:

1) Download my fork
2) switch to *one point fifteen point two* branch
3) build and move `dist/baritone-api-1.5.3.jar` to `.minecraft/libraries/cabaletta/baritone-api/1.5.3`
4) it will work

**Comment by ZacSharp** [comment]
1. Done
2. Done by accident
3. Uh, I'm always using forge or a dev env, so I have to set up a new install for that
4. We'll see, didn't work from dev env

**Comment by mvrozanti** [comment]
I edited my previous answer in case that doesn't work.

**Comment by ZacSharp** [comment]
If you really want to you can send me a precompiled jar, but getting identical files from different places won't help.

**Comment by mvrozanti** [comment]
Please, never run a precompiled jar of unknown origin. 

You say you have the latest version, but this error:


> Failed to find block id for minecraft:air

Says otherwise.

**Comment by ZacSharp** [comment]
That error is gone but the screenshots are still transparent. Haven't tested a normal install yet because the game crashes when I try to start it.

**Comment by mvrozanti** [comment]
Do you have discord or something?

**Comment by ZacSharp** [comment]
No

**Comment by mvrozanti** [comment]
Well, at the moment I can only tell you to do the necessary steps for getting the branch in this PR, which is the version that generated the picture I posted above.

**Comment by ZacSharp** [comment]
Ok, so apparently I can't test this in a production enviroment because the baritone json version crashes and the forge jar is generated by the build step that just crashes. I'll now go and try to get this working without a working start. Thank you for trying to help me.

**Comment by mvrozanti** [comment]
Thank **you** for trying to help us get this feature

**Comment by ZacSharp** [comment]
Just a few things from the last 20 minutes:
1. How do I find out if a region is empty?
2. Player873 is banned on one of my two testing servers (on topic I know)
3. ```java
     private BufferedImage getTextureForBlockId(String blockId) throws IOException {
         String base = "/home/mcbw/.minecraft/versions/1.12.2/1.12.2/assets/minecraft/textures/blocks/";
         int x = 0/0;
         System.out.println(blockId + " -> " + String.format("%s",x));
        ...
   ```
does not cause any exception or output
4. This causes weird errors when switching server without restarting

1 and 3 are most important rn

**Comment by mvrozanti** [comment]
1) I consider a region to be empty if every single block in that region (which means the blocks from northwest corner to southwest corner in all y levels) returns null.

2) I have no idea what you're talking about

3) I'm not sure what you're trying to achieve with that piece of code

4) I haven't tested switching servers without restarting

**Comment by ZacSharp** [comment]
2. Not important, just some random username being banned on some random cracked server
3. I tried to get some output to see if and when it is executed. The x=0/0 was would have caused a RuntimeException, if it had been executed.

**Comment by mvrozanti** [comment]
That's a weird way to test but ok. You need to make that exception pop up somehow... Have you followed the instructions on SETUP.md?

**Comment by ZacSharp** [comment]
Jup, the ones for command line.

**Comment by mvrozanti** [comment]
I'm going to assume "Jup" means yes. 

But you also need to follow the [More Info](https://github.com/cabaletta/baritone/blob/master/SETUP.md#more-info) section.

**Comment by ZacSharp** [comment]
Your assumption is correct.
That's a bit hard to do at a restart frequency of one per 10 to 20 seconds.

**Comment by mvrozanti** [comment]
>  That's a bit hard to do at a restart frequency of one per 10 to 20 seconds.

What is?

**Comment by ZacSharp** [comment]
My current workflow: runClient, wait, join server, #tomap, alt + F4, alt + tab, little code change, alt + tab, repeat
And copying over files every time is really annoying

**Comment by mvrozanti** [comment]
**Follow the More Info section** [here](https://github.com/cabaletta/baritone/blob/master/SETUP.md#more-info).

**Comment by ZacSharp** [comment]
That's the instructions for users using impact. I'm a dev without impact, so I look at build it yourself.

**Comment by mvrozanti** [comment]
Ok that makes sense then. I'm not sure why your client isn't reflecting the code changes though. I think you need to work on that first.

**Comment by ZacSharp** [comment]
It is. e.g. when I put in output before the double for loop that gets sent

**Comment by mvrozanti** [comment]
What are you having trouble with then?

**Comment by ZacSharp** [comment]
I was just telling you that the code seems to not be run, maybe you can check if you get output from there.

**Comment by mvrozanti** [comment]
Can you elaborate on that?

**Comment by ZacSharp** [comment]
some example values (with applied offsets)
minX = 0
maX = 511
minZ = 2147483647
maxZ 1

**Comment by mvrozanti** [comment]
minX, maxX, minZ, maxZ should result in the (block) coordinates of the northwest and southeast corners of the region, respectively. I'm about to go offline but I'm gonna check this part of the code asap

**Comment by ZacSharp** [comment]
are the coordinates world coordinates or 0-511?
ok, then until tomorrow

**Comment by mvrozanti** [comment]
> are the coordinates world coordinates or 0-511?

Please be more descriptive in your questions.

**Comment by ZacSharp** [comment]
what is the range of mixX/maxX/minZ/maxZ? Are they relative to the world origin or to the region origin?

**Comment by mvrozanti** [comment]
I can only say that `maxX-minX=512` and `maxZ-minZ=512` in **block coordinates**, which are different than chunk coordinates and region coordinates.

**Comment by ZacSharp** [comment]
ok, imma go off for today

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+                                }
+                            } catch (Exception e) {
+//                            System.out.println("Couldn't get BlockState for " + x + " " + y + " " + z);
+                            }
+                        }
+                        if (!topView.containsKey(x)) {
+                            if (!topView.get(x).containsKey(z)) {
+                                topView.get(x).put(z, DEFAULT_BLOCK_ID); // this is definitely not accurate
+                            }
+                        }
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:186]: Here's the culprit: You are only trying to insert missing z-entries when the whole x-row is missing, which is guaranteed to be false.
```suggestion
                        if (!topView.get(x).containsKey(z)) {
                            topView.get(x).put(z, DEFAULT_BLOCK_ID); // this is definitely not accurate
                        }
```

Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:186]: Damn. Sorry, I'm not following but I have a feeling you found it problem

Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:186]: This does not fix the missing block states but it does fix the loop range. Basically whenever we don't get a blockstate that entry is missing in topView.get(x) and the code you had there wasn't fixing this, so when we later use the entries to find the minimum z position there is no key lower that Interger.MAX_VALUE, because there is no key at all.

File: src/main/java/baritone/command/defaults/ToMapCommand.java [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+                logDirect("Getting top blocks for region " + rX + " " + rZ);
+                for (int x = nw[0]; x < nw[0] + 512; x++) {
+                    if (!topView.containsKey(x)) {
+                        HashMap<Integer, String> zId = new LinkedHashMap<>();
+                        topView.put(x, zId);
+                    }
+                    for (int z = nw[1]; z < nw[1] + 512; z++) {
+                        for (int y = 256; y >= 0; y--) {
+                            try {
+                                BlockState bs = cr.getBlock(x, y, z);
```
Comment (ZacSharp) [inline src/main/java/baritone/command/defaults/ToMapCommand.java:154]: this uses coordinates relative to the region, as shown in `BlockStateInterface#get0`
Swallowing exceptions seems to be a bad idea, this one causes up to 512*512*256 OutOfBounds exceptions per region and no one noticed it. Now I'm getting actual null blockstates.

Comment (mvrozanti) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:154]: I don't have the time to verify this now. Are you implying that there's a problem in baritone's code itself or is it this problem restricted to ToMapCommand? Also, do you think this is easily fixable?

Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/ToMapCommand.java:154]: Actuall I don't know. This might be a problem of ToMapCommand not accessing the cache correctly or baritone not returning blockstates it is supposed to return. I will try to use the normal ctx.get access method now.

**Comment by mvrozanti** [comment]
@ZacSharp I haven't had the time to look into it but @FriedrichM is also lending us a hand. If can [test it](https://github.com/FriedrichM/baritone/commit/17d8cebd575960c9419e192e8c408188e765ec32)

**Comment by ZacSharp** [comment]
Worked first time. Is not singleplayer safe (easy fix, already did it for testing). Maps don't look really well (as expected). No need for unzipped assets but coloring only looks for the material, resulting in a small colour palette (difficult to fix as the material colour is the only colour directly available from block states I know). Doesn't freeze the game (easy if think).

I managed to get something working yesterday as well, but the heightMap calculation seem to be wrong. If you are interested you can find it in my fork (I just did a quick commit to save it, so it contains all testing changes I had done at that point).

**Comment by mvrozanti** [comment]
I just tried translating it to 1.15.2 and I'm getting blank (transparent) images as well...

**Comment by mvrozanti** [comment]
This is great, thanks for the help @FriedrichM. Turns out it wasn't a bug in baritone then.

**Comment by ZacSharp** [comment]
is it possible to change the branch a pr is merged into? if so please switch this pr to master.

**Comment by l1ving** [comment]
No, it's not.

**Comment by mvrozanti** [comment]
image

Can we get rid of the red lines though? How are they coming into view? I assume those are chunk borders; does IBlockState#getMaterial() change in this case? Why only on water?

**Comment by l1ving** [comment]
Ooh that's cool

**Comment by FriedrichM** [comment]
i am puzzled how these red lines arise. it never happend in my tests, maybe these are old cache files and they changed the format at some point. ill have a look at it and ill also look at how to efficiently stitch those images together

**Comment by FriedrichM** [comment]
While the colormapping is in fact off (as mentioned in my code) i could not recreate the red lines even with old cache files. 

Stitching images together will be hard without using much ram and it will especially be hard without the use of a library. 
I doubt it will be worth adding a library to the jar just for this command so my suggestion is to use imagej [https://imagej.net/Fiji](https://imagej.net/Fiji) with the preinstalled stitching plugin and let the command generate a TileConfiguration.txt file. 
This file can then be used with imagej to stitch the files together.
The downsides are that imagej can be a bit hard to use and imagej doesnt seem to work with too tiled maps, say worldborder to spawn.
I looked into the code (https://github.com/fiji/Stitching) but i couldnt find the reason on first sight, i think its some variable overflowing to the negative tho because I get a negative arrayindex error.

Another suggestion would be to make an option to render all cached regions in a radius around a given coordinate, this would be more flexible than to render all of them.

**Comment by mvrozanti** [comment]
As per the red lines appearing in my example: I have a mild suspicion that might be happening because of another mod of mine that represents new chunks with these red lines. I'm gonna test without it on later today

**Comment by ZacSharp** [comment]
I never had such a problem with [this](https://github.com/ZacSharp/baritone/blob/tomap-command-from-master/src/main/java/baritone/command/defaults/ToMapCommand.java). Would be cool if you actually merged this into your tomap-from-master branch because it was a lot of work and it contains everything @FriedrichM did (merged his branch into mine).

**Comment by FriedrichM** [comment]
add returns to the "Failed!" printouts

**Comment by ZacSharp** [comment]
I thought about it and then left it the way you did it, so the user (and probably maintainers getting it as a bug) gets a proper exception.
It's easier like that but I can change it to `e.printStackTrace(); return;`

**Comment by ZacSharp** [comment]
Is it worth it trying to do something like in [MapWriter](https://github.com/daveyliam/mapwriter/blob/master/src/main/java/mapwriter/BlockColourGen.java) or is that overcomplicated?

**Comment by mvrozanti** [comment]
Seems like we now have 3 different ways of getting color of blocks. 

But we can also choose that later.

The stitching or tiling part now seems to be the most difficult challenge, especially from a memory efficiency aspect. These maps get out of hand really fast.

**Comment by ZacSharp** [comment]
I think it is possible to do it with bitmaps, if you really want to.
The downsides are:
1. Really slow because we reload every region from disk 512 times and write to disk a lot.
2. Takes more space on disk because the image is uncompressed.
I don't know if this would be possible with compressed formats but it will always stay slow.

**Comment by ZacSharp** [comment]
I've got a stitched image here, but apparently it is either to big or the wrong file type for github to render it.
image
How many regions do you actually intend to stitch? Because you can keep up to 1271 (and a half) region images in 1GB ram.

**Comment by mvrozanti** [comment]
Maybe Github doesn't like .ppms :smile: 

> How many regions do you actually intend to stitch? Because you can keep up to 1271 (and a half) region images in 1GB ram.

I'd say as many as possible.

**Comment by ZacSharp** [comment]
So that approach keeps a maximum of `512 pixels * 3 byte per pixel = 1537` bytes in ram, for any output size. I guess that means as long as your hard drive can handle 10,8PB (1,08×10¹⁶ byte) you can stitch the whole world using this.

**Comment by l1ving** [comment]
> I've got a stitched image here, but apparently it is either to big or the wrong file type for github to render it.

Yea, github refuses to render it if it's too big. I learned that the hard way with gifs
